### PR TITLE
#2057: Shuffle outgoing transport parameters.

### PR DESCRIPTION
**Pull request by mstyura**
I propose to remember the order in which write of the params should be performed when `TransportParameters` are crated. Then this order is used to write to wire format.
PR currently split into two commits:
1. Do the permutation using hard-coded IDs of prams;
2. Some refactoring to store the IDs into enum.

### CODE REVIEW

File: quinn-proto/src/transport_parameters.rs
Comment (Ralith) [inline quinn-proto/src/transport_parameters.rs:73]: Factor the `TransportParameterId` qualifier out into the macro for readability.

File: quinn-proto/src/transport_parameters.rs
Comment (Ralith) [inline quinn-proto/src/transport_parameters.rs:379]: style: prefer `id as u64` here and below so there's less risk of copy-paste errors in the future.

File: quinn-proto/src/transport_parameters.rs
Comment (Ralith) [inline quinn-proto/src/transport_parameters.rs:60]: nit: This seems much more verbose than necessary.
```suggestion
    MinAckDelayDraft07 = 0xFF04DE1B,
```

**Comment by djc** [comment]
> I propose to remember the order in which write of the params should be performed when `TransportParameters` are crated. Then this order is used to write to wire format.

Why do you want to store the order? That seems largely unrelated to the stated goal of this PR to shuffle transport params.

I think the commit introducing an enum should come first. Suggest extracting the macro from its current location in a separate commit so that it can get only the necessary modifications in place in the further commits.

**Comment by mstyura** [comment]
> Why do you want to store the order? That seems largely unrelated to the stated goal of this PR to shuffle transport params.

@djc maybe I'm missing something, but inside `TransportParameters::write` I don't have access to random generator passed form `Connection` to perform shuffle at that time and in my previous MR it was suggested to not to use `rand::thread_rng()` directly. So the solution was to store the write order inside constructor where I have an access to random generator injected from `Connection` and use this order during `write`. Let me know if you think this could be done differently.

> I think the commit introducing an enum should come first. Suggest extracting the macro from its current location in a separate commit so that it can get only the necessary modifications in place in the further commits.

Do you mean only re-order commits in this MR or I should better to put refactoring related to `TransportParameterId` into separate PR and wait until it merged and update this MR with only shuffling implemented?

**Comment by djc** [comment]
> > Why do you want to store the order? That seems largely unrelated to the stated goal of this PR to shuffle transport params.
> 
> @djc maybe I'm missing something, but inside `TransportParameters::write` I don't have access to random generator passed form `Connection` to perform shuffle at that time and in my previous MR it was suggested to not to use `rand::thread_rng()` directly. So the solution was to store the write order inside constructor where I have an access to random generator injected from `Connection` and use this order during `write`. Let me know if you think this could be done differently.

Ah, makes sense -- sorry for the noise!

> > I think the commit introducing an enum should come first. Suggest extracting the macro from its current location in a separate commit so that it can get only the necessary modifications in place in the further commits.
> 
> Do you mean only re-order commits in this MR or I should better to put refactoring related to `TransportParameterId` into separate PR and wait until it merged and update this MR with only shuffling implemented?

Reordering commits in this PR should be okay.

**Comment by mstyura** [comment]
> Reordering commits in this PR should be okay.

Re-ordered commits and hopefully a bit enhanced implementation by allowing `==` between `TransportParameterId` and `u64` to avoid several usages `as u64` casts inside guard of match expression.

> Ah, makes sense -- sorry for the noise!

Also a bit changed implementation of reordering. Instead of storing all transport parameter IDs inside `write_order` array inside `TransportPrameters` which occupied `21 * u64 = 168` bytes spaces in struct I stored `21 * u8` values which treated as indices of `SUPPORTED_TRANSPORT_PARAMETERS` array.

File: quinn-proto/src/transport_parameters.rs
Comment (mstyura) [reply quinn-proto/src/transport_parameters.rs:175]: Not really mandatory, I though `write_order` only make sense for instances of `TransportParameters` which are created to be then send, so received `TransportParameters` has `None` as `write_order`. I can change code to avoid `Option`, should I?

Comment (djc) [reply quinn-proto/src/transport_parameters.rs]: Fair -- feels a little ugly that we're sticking it into `TransportParameters` but I see your point that it's not trivial to pass the RNG around to where we need it. Can you at least more clearly document on the field why it is `Option` and when it is expected to be `Some` vs `None`? (Keep the current comment on its own line, then an empty line, then the additional context.)

File: quinn-proto/src/transport_parameters.rs
Comment (djc) [inline quinn-proto/src/transport_parameters.rs:434]: One more thing! Could we do a `TryFrom<u64> for `TransportParameterId` impl with `Err = ()` so that we can simplify these match arms?

**Review by Ralith — APPROVED** [review]
Thanks!

**Review by djc — APPROVED** [review]

**Comment by mstyura** [comment]
@djc, @Ralith thank you very much for review, feedback and patience! And accepting this change!

**Comment by Ralith** [comment]
Thanks for your contribution! I've been specifically wanting this for a while.

**Pull request merged**
